release_util/management/commands/__init__.py

```diff
@@ -132,164 +134,215 @@ def state(self):
+        states for apps with unapplied migrations.
 
-            # Find the apps that have been applied -or- failed.
-            # Include their initial migrations as commands.
-            self._add_rollback_commands()
-            raise CommandError("Migration failed for app '{}' - migration '{}'.\n".format(failed_app, failed_migration))
+        Both are tuples of the form (app, migration_name).
+        """
+        connection = connections[self._database_name]
+        loader = MigrationLoader(connection, ignore_no_migrations=True)
+        unapplied = self._gather_unapplied_migrations(loader)
+        currents = self._gather_current_migration_state(loader, [u[0] for u in unapplied])
+        return unapplied, currents
 
-        # All migrations succeeded. Record the total time for all migrations.
-        time_to_apply = self.timer() - start
-        self.migration_state['duration'] = time_to_apply
+    def _parse_migrate_output(self, output):
+        """
+        Returns (succeeded:list(tuple), failed:tuple or None)
+        Both tuples are of the form (app, migration)
+        """
+        failed = None
+        succeeded = []
 
-        # Parse the output to find all migrations that were applied successfully.
-        for line in out.getvalue().split('\n'):
+        # Mark migrations:
+        # - before exception migration as success
+        # - exception migration as failed
+        # - any remaining exceptions as unapplied
+        for line in output.split('\n'):
             line = _remove_escape_characters(line).strip()
             line_match = self.migration_regex.match(line)
             if line_match:
-                migration = [line_match.group('app_name'), line_match.group('migration_name')]
+                migration = (line_match.group('app_name'), line_match.group('migration_name'))
                 if line_match.group('success') == 'OK':
-                    # A migration succeeded.
-                    self.migration_state['success'].append({
-                        'migration': migration,
-                        'output': line,
-                    })
+                    # The migration succeeded
+                    succeeded.append(migration)
+                else:
+                    # The migration failed
+                    failed = migration
+                    break
+        return succeeded, failed
 
-    def apply_all(self):
+    def _apply(self, migration=None, run_all=False):
+        """ 
+        If a migration is supplied, runs that migration and appends to state.
+        If run_all==True, runs all migrations.
+        Raises a ValueError if neither "migration" nor "run_all" are provided.
+        """
+        out = StringIO()
+        failed = traceback = None
+        
+        migrate_kwargs = {
+            'interactive': False,
+            'stdout': out,
+            'database': self._database_name,
+        }
+        if migration is not None:
```

Comment:
if a user specifies a migration and says run_all, is this the expected behavior?